dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -311,18 +311,34 @@ ExchangeReceiverBase<RPCContext>::ExchangeReceiverBase(
 template <typename RPCContext>
 ExchangeReceiverBase<RPCContext>::~ExchangeReceiverBase()
 {
-    setState(ExchangeReceiverState::CLOSED);
-    msg_channel.finish();
+    close();
     thread_manager->wait();
 }
 
 template <typename RPCContext>
 void ExchangeReceiverBase<RPCContext>::cancel()
 {
+    auto cur_state = getState();
+    if (cur_state == ExchangeReceiverState::CANCELED || cur_state == ExchangeReceiverState::CLOSED)
+    {
+        return;
+    }
     setState(ExchangeReceiverState::CANCELED);
     msg_channel.finish();
 }
 
+template <typename RPCContext>
+void ExchangeReceiverBase<RPCContext>::close()
+{
+    auto cur_state = getState();
+    if (cur_state == ExchangeReceiverState::CANCELED || cur_state == ExchangeReceiverState::CLOSED)
+    {
+        return;
+    }
```

Comment:
this check should be moved into `setState` and under protection of `mu`.

Reply:
`getState` is protected by `mu`